tests/testthat/test-method-dispatch.R

```diff
@@ -225,5 +225,21 @@ test_that("method dispatch works for class_missing", {
     variant = if (getRversion() < "4.3") "R-lt-4-3",
     foo_wrapper()
   )
+})
+
+test_that("errors from dispatched methods have reasonable tracebacks", {
+  my_generic <- new_generic("my_generic", "x")
```

Comment:
I think you could do something like this instead:

```R
my_generic <- new_generic("my_generic", "x")
method(my_generic, class_double) <- function(x) sys.call(-3)
my_generic(10)
```

(I found the magic number of -3 through trial and error 😬)

Reply:
If you instead use `rlang::trace_back()`:

```R
my_generic <- new_generic("my_generic", "x")
method(my_generic, class_double) <- function(x) rlang::trace_back()
my_generic(10)
```

You then see the full tree:

```
▆
 1. └─global my_generic(10)
 2.   ├─S7::S7_dispatch()
 3.   │ └─base::eval(S7_dispatched_call, envir = sys.frame(-1)) at OOP-WG/R/method-dispatch.R:24:3
 4.   │   └─base::eval(S7_dispatched_call, envir = sys.frame(-1))
 5.   └─global `<S7_methd>`(`<dbl>`, ... = ...)
 ```

Since we're in here, that makes me wonder if we should eliminate the two eval calls (which both come from `base::eval()`) by doing the evaluation in C. i.e. instead of doing this part in C and part in R:

```R
S7_dispatch <- function() {
  S7_dispatched_call <- .Call(method_call_, sys.call(-1), sys.function(-1), sys.frame(-1))
  eval(S7_dispatched_call, envir = sys.frame(-1))
}
```

Should we implement `S7_dispatch()` purely in C?